dotlockfile.1
.TH DOTLOCKFILE 1 "January 10, 2017" "" "Cistron Utilities"
.SH NAME
dotlockfile \- Utility to manage lockfiles
.SH SYNOPSIS
.B dotlockfile
.B \-l
.RB [ \-r
.IR retries ]
.RB [ \-i
.IR interval ]
.RB [ \-p ]
.RB [ \-q ]
.RB < \-m \ |
.IR lockfile >
.br
.B dotlockfile
.B \-l
.RB [ \-r
.IR retries ]
.RB [ \-i
.IR interval ]
.RB [ \-p ]
.RB [ \-q ]
.RB < \-m \ |
.IR lockfile >
.RB [ \-P ]
.IR cmd "\ args \&...\&"
.br
.B dotlockfile
.RB \-u \ | \ \-t
.br
.SH DESCRIPTION
.B dotlockfile
is a command line utility to reliably create, test and remove lockfiles.
It creates lockfiles
.I reliably
on local and NFS filesystems, because the crucial steps of testing for a
preexisting lockfile and creating it are performed
.I atomically
by a
.I single
call to
.IR link (2).
Manpage
.IR lockfile_create (3)
describes the used algorithm.
.PP
.B dotlockfile
is installed with attribute
.BI SETGID \ mail
and thus can also be used to lock and unlock mailboxes
.I even
if the mailspool directory is only writable by group mail.
.PP
The name
.B dotlockfile
comes from the way mailboxes are locked for updates on a lot of UNIX systems.
A lockfile is created with the same filename as the mailbox but with the string
"\fI.lock\fR" appended.
.PP
The names
.B dotlock
and
.B lockfile
were already taken \(en hence the name \fBdotlockfile\fR \fI:)\fR.
.SH OPTIONS
.IP "\fB\-l\fR"
Create a lockfile if no preexisting valid lockfile is found, else wait and retry
according to option \fB\-r\fR.  Retry interval can be explicitly set with option \fB\-i\fR.
This option (\fB-l\fR) is the default, so it can be left off.

A lockfile is treated as valid,
.br
\[bu]\ \ if it holds the
.I process\-id
of a running process,
.br
\[bu]\ \ or if it does not hold any
.I process\-id
and has been touched less than 5\ minutes ago (timestamp is younger than
5\ minutes).
.IP "\fB\-r retries\fR"
The number of times
.B dotlockfile
retries to acquire the lock if it failed the first time before giving up.
The initial sleep after failing to acquire the lock is 5\ seconds.
After each retry the sleep interval is increased incrementally by 5\ seconds
up to a maximum sleep of 60\ seconds between tries unless overridden by \fB\-i\fR.
The default number of retries is 5.
To try only once, use "\fB\-r 0\fR".
To try indefinitely, use "\fB\-r \-1\fR".
.IP "\fB\-i interval\fR"
Sets a consistent retry interval.
.IP "\fB\-u\fR"
Remove a lockfile.
.IP "\fB\-t\fR"
Touch an existing lockfile (update the timestamp).
Useful for lockfiles on NFS filesystems.
For lockfiles on local filesystems the
.B \-p
option is preferable.
.IP "\fB\-p\fR"
Write the
.I process\-id
of the calling process (or dotlockfile itself if a command is executed)
into the lockfile.
Also when testing for an existing lockfile, check the contents for the
.I process\-id
of a running process to verify if the lockfile is still valid.
Obviously useful only for lockfiles on local filesystems.
.IP "\fB\-m\fR"
Lock or unlock the current users mailbox.
The path to the mailbox is the default system mailspool directory (usually
.IR /var/mail )
with the username as gotten from
.IR getpwuid ()
appended.
If the environment variable
.I $MAIL
is set, that is used instead.
Then the string "\fI.lock\fR" is appended to get the name of the actual
lockfile.
.IP "\fB\-q\fR"
Don't print warnings or errors to the standard error output. Used internally
by liblockfile when it spawns
.B dotlockfile
as a helper program.
.IP "\fB\-P\fR"
On successful "lock and spawn command", don't exit with status zero, but
pass through the exit value of the spawned command.
.IP lockfile
The lockfile to be created or removed.
Must not be specified if the \fB\-m\fR option is given.
.IP "\fIcommand\fR argument \&...\&"
Create lockfile, run the
.I command
, wait for it to exit, and remove lockfile.
.SH RETURN\ VALUE
Zero on success, and non\-zero on failure.
When locking (the default, or with the \fB\-l\fR option)
.B dotlockfile
returns the same values as the library function
.IR lockfile_create (3).
Unlocking a non\-existent lockfile is not an error.
.PP
Unless the
.B \-P
option was supplied, when a command is executed, the return value does not
correspond with that of the command that was run.  If locking and unlocking
was successful, the exit status is zero.
.SH NOTES
The lockfile is created exactly as named on the command line.
The extension "\fI.lock\fR" is \fInot\fR automatically appended.
.PP
This utility is a lot like the
.IR lockfile (1)
utility included with
.IR procmail ,
and the
.IR mutt_dotlock (1)
utility included with
.IR mutt .
However the command\-line arguments differ, and so does the return status.
It is believed, that
.I dotlockfile
is the most flexible implementation, since it automatically detects when it
needs to use privileges to lock a mailbox, and does it safely.
.PP
The above mentioned
.IR lockfile_create (3)
manpage is present in the
.I liblockfile\-dev
package.
.SH BUGS
None known.
.SH SEE\ ALSO
.IR lockfile_create (3),
.IR maillock (3)
.SH AUTHOR
Miquel van Smoorenburg


Changelog
liblockfile (1.17)

  * dotlockfile: '-P' option: allow passing through the child exit code
  * dotlockfile: '-i interval' option.
  * dotlockfile: document -q
  * add L_RMSTALE return status (failed to remove stale lock)
  * remove cistron.nl email address

 -- Miquel van Smoorenburg Wed, 27 Jan 2021 00:33:10 +0100

    v1.16

liblockfile (1.16)

  * lockfile.c: remove all the permission checking in the normal case.
    Only if creating the tempfile / removing the lockfile returns EACCESS,
    check to see if this is a mail-lockfile, if so launch helper.
  * dotlockfile.c: use same mail-lockfile check as lockfile.c, but a
    bit more strict.
  * dotlockfile: actually honor -q flag (quiet) everywhere

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Mon, 29 Jul 2019 19:31:09 +0200

liblockfile (1.15)

  * dotlockfile.1: Fix groff syntax (from 1.14-1.1 debian NMU by Mattia Rizzolo <mattia@debian.org>)
  * dotlockfile.1: Some minor fixes in the manual (from https://bugs.debian.org/899344)
  * fix getgroups() check in lockfile.c (Thomas Finke), add some cleanups
  * fix gcc warning about snprintf in lockfile_create_save_tmplock() (not an actual bug)
  * fix gcc warning about strncpy in ismaillock() (not an actual bug)
  * add support for systems that do not have PATH_MAX
    (fixes https://github.com/miquels/liblockfile/issues/1)
  * rename configure.in to configure.ac and run autoconf
    (fixes https://github.com/miquels/liblockfile/issues/2)

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Tue, 23 Jul 2019 23:25:53 +0200

liblockfile (1.14)

  * allow '-m' option again when executing command (why disallow it)
  * drop privs just after fork (and before execv) instead of earlier
  * before executing command, chdir back to original working directory.

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Tue, 17 Jan 2017 13:58:25 +0100

liblockfile (1.13)

  * remove '-t' as option when running a command- do touchlock
    automatically if -p was not given.
  * drop effective gid asap, retain as saved gid. only restore egid if nessecary.
  * if stat() fails, drop saved gid as well, and retry (sshfs is strict)
  * don't allow executing a command with the -m option, not nessecary.
    should really add a minimal setgid util for maillock(), remove setgid
    from dotlockfile, and let it use the library maillock().

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Thu, 05 Jan 2017 23:10:35 +0100

liblockfile (1.12)

  * update lockfile_create.3 manpage, document L_PPID

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Thu, 05 Jan 2017 11:37:22 +0100

liblockfile (1.11)

  * fix dotlockfile -r -1 functionality
  * add dotlockfile <lockfile> cmd args... functionality
  * dotlockfile: use geteuid() instead of getuid() to find out mailboxname.
  * before forking an external program, first check geteuid() != 0,
    then do setuid(geteuid())

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Tue, 03 Jan 2017 20:45:15 +0100

liblockfile (1.10)

  * add complete GPL/LGPL licenses to source code
  * Merge all debian patches into upstream
  * fail to lock if parent is PID 1 - makes no sense to write PID 1
    as owning process into the lockfile (Per Cederqvist <cederp@opera.com>)

 -- Miquel van Smoorenburg <mike@langeraar.net>  Fri, 18 Nov 2016 14:14:28 +0100

liblockfile (1.09)

  * Update path to full license in the COPYRIGHT file
    (Anibal Monsalve Salazar)
  * Remove debug code from lockfile_create():
     The lockfile_create-function still contains some debug code, which
     considers a lock stale after 10 seconds instead of 5 minutes. As
     the documentation suggests refreshing the lock every 60 seconds,
     programs might lose the lock unexpectedly. Found while researching
     debian bug #505278. (Wolfram Sang <w.sang@pengutronix.de>)
  * Updated dotlockfile manpage and updated dotlockfile usage message.
    Fixes debian bug 493462 (Roland Eggner <edvd1@systemanalysen.net>)
  * Change 'ar' in Makefile.in to '$(AR)' (for cross builds).
    Fixes debian bug 562937 (Wolfram Sang <wsa@pengutronix.de>)
  * Clean up temporary lockfile if lockfile(1) was interrupted.
  * dotlockfile now also allows you to create a lockfile in
    a group-writable-by-one-of-your-auxilary-groups directory
  * dotlockfile: -r-1 now retries indefinitely (well, 4000 years)
  * remove ppsh macro in dotlockfile.1, replace with standard .SH
    (Iain Calder <dn097@torfree.net>)

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Wed, 20 Jul 2011 15:54:58 +0200

liblockfile (1.08)

  * When not writing a pid into the lockfile, write "0\n" as
    placeholder instead of "0\0" (bug reported by Rob Browning
    <rlb@defaultvalue.org>)
  * lockfile_check() did the following:
      - open lockfile
      - read from it
      - stat file and get atime as 'now'
    this gets the 'now' time from the filesystem so comparisations
    to mtime work even if the clock of the remote file server is off.
    However this doesn't work with noatime / relatime, so now we stat
    the lockfile both before and after the read and use atime only
    if it differs before and after.
  * Copy Changelog from debian/changelog now that the package
    isn't native anymore since 1.07

 -- Miquel van Smoorenburg <miquels@cistron.nl>  Thu, 24 Jul 2008 22:40:34 +0200

liblockfile (1.07-1) unstable; urgency=low

  * Fixed "liblockfile1: option -u is duplicated". Closes: #416271

 -- Anibal Monsalve Salazar <anibal@debian.org>  Sat, 05 Apr 2008 12:14:10 +1000

liblockfile (1.06.2) unstable; urgency=low

  * Non-maintainer upload.
  * Rename eaccess() to eaccess_write() as eaccess is now a glibc function.
    Patch by Colin Watson. Closes: #392779.

 -- Aurelien Jarno <aurel32@debian.org>  Sun, 03 Jun 2007 19:28:33 +0200

liblockfile (1.06.1) unstable; urgency=low

  * Non-maintainer upload.
  * Remove postinst and prerm scripts, completing the /usr/doc tarnsition.
    Closes: #322790

 -- Joey Hess <joeyh@debian.org>  Tue, 10 Jan 2006 01:21:31 -0500


dotlockfile.c
/*
 * dotlockfile.c	Command line version of liblockfile.
 *			Runs setgid mail so is able to lock mailboxes
 *			as well. Liblockfile can call this command.
 *
 *		Copyright (C) Miquel van Smoorenburg and contributors 1999-2021
 *
 *		This program is free software; you can redistribute it and/or
 *		modify it under the terms of the GNU General Public License
 *		as published by the Free Software Foundation; either version 2
 *		of the License, or (at your option) any later version.
 */

#include "autoconf.h"

#include <sys/types.h>
#if HAVE_SYS_PARAM_H
#include <sys/param.h>
#endif
#include <sys/wait.h>
#include <stdio.h>
#include <string.h>
#include <pwd.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <time.h>
#include <errno.h>
#include <maillock.h>
#include <lockfile.h>

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#ifndef HAVE_GETOPT_H
extern int getopt();
extern char *optarg;
extern int optind;
#endif

extern int is_maillock(const char *lockfile);
extern int lockfile_create_set_tmplock(const char *lockfile,
			volatile char **tmplock, int retries, int flags, struct lockargs *);

static volatile char *tmplock;
static int quiet;

/*
 *	If we got SIGINT, SIGQUIT, SIGHUP, remove the
 *	tempfile and re-raise the signal.
 */
void got_signal(int sig)
{
	if (tmplock && tmplock[0])
		unlink((char *)tmplock);
	signal(sig, SIG_DFL);
	raise(sig);
}

void ignore_signal(int sig)
{
}

/*
 *	Install signal handler only if the signal was
 *	not ignored already.
 */
int set_signal(int sig, void (*handler)(int))
{
	struct sigaction sa;

	if (sigaction(sig, NULL, &sa) < 0)
		return -1;
	if (sa.sa_handler == SIG_IGN)
		return 0;
	memset(&sa, 0, sizeof(sa));
	sa.sa_handler = handler;
	return sigaction(sig, &sa, NULL);
}

/*
 *	Sleep for an amount of time while regulary checking if
 *	our parent is still alive.
 */
int check_sleep(int sleeptime, int flags)
{
	int		i;
	int		interval = 5;
	static int	ppid = 0;

	if (ppid == 0) ppid = getppid();

	if (flags & L_INTERVAL)
		interval = 1;

	for (i = 0; i < sleeptime; i += interval) {
		sleep(interval);
		if (kill(ppid, 0) < 0 && errno == ESRCH)
			return L_ERROR;
	}
	return 0;
}

/*
 *	Split a filename up in  file and directory.
 */
int fn_split(char *fn, char **fn_p, char **dir_p)
{
	static char	*buf = NULL;
	char		*p;

	if (buf)
		free (buf);
	buf = (char *) malloc (strlen (fn) + 1);
	if (! buf)
		return L_ERROR;
	strcpy(buf, fn);
	if ((p = strrchr(buf, '/')) != NULL) {
		*p++   = 0;
		*fn_p  = p;
		*dir_p = buf;
	} else {
		*fn_p  = fn;
		*dir_p = ".";
	}
	return L_SUCCESS;
}


/*
 *	Return name of lockfile for mail.
 */
char *mlockname(char *user)
{
	static char	*buf = NULL;
	char		*e;

	if (buf)
		free(buf);

	e = getenv("MAIL");
	if (e) {
		buf = (char *)malloc(strlen(e)+6);
		if (!buf)
			return NULL;
		sprintf(buf, "%s.lock", e);
	} else {
		buf = (char *)malloc(strlen(MAILDIR)+strlen(user)+6);
		if (!buf)
			return NULL;
		sprintf(buf, "%s%s.lock", MAILDIR, user);
	}
	return buf;
}

void perror_exit(const char *why) {
	if (!quiet) {
		fprintf(stderr, "dotlockfile: ");
		perror(why);
	}
	exit(L_ERROR);
}

/*
 *	Print usage mesage and exit.
 */
void usage(void)
{
	fprintf(stderr, "Usage:  dotlockfile -l [-r retries] [-i interval] [-p] [-q] <-m|lockfile>\n");
	fprintf(stderr, "        dotlockfile -l [-r retries] [-i interval] [-p] [-q] <-m|lockfile> [-P] command args...\n");
	fprintf(stderr, "        dotlockfile -u|-t\n");
	exit(1);
}

int main(int argc, char **argv)
{
	struct passwd	*pwd;
	struct lockargs args;
	gid_t		gid, egid;
	char		*lockfile = NULL;
	char		**cmd = NULL;
	int 		c, r;
	int		retries = 5;
	int		interval = 0;
	int		flags = 0;
	int		lock = 0;
	int		unlock = 0;
	int		check = 0;
	int		touch = 0;
	int		writepid = 0;
	int		passthrough = 0;

	/*
	 *	Remember real and effective gid, and
	 *	drop privs for now.
	 */
	if ((gid = getgid()) < 0)
		perror_exit("getgid");
	if ((egid = getegid()) < 0)
		perror_exit("getegid");
	if (gid != egid) {
		if (setregid(-1, gid) < 0)
			perror_exit("setregid(-1, gid)");
	}

	memset(&args, 0, sizeof(struct lockargs));

	set_signal(SIGINT, got_signal);
	set_signal(SIGQUIT, got_signal);
	set_signal(SIGHUP, got_signal);
	set_signal(SIGTERM, got_signal);
	set_signal(SIGPIPE, got_signal);

	/*
	 *	Process the options.
	 */
	while ((c = getopt(argc, argv, "+qpNr:mluci:tP")) != EOF) switch(c) {
		case 'q':
			quiet = 1;
			break;
		case 'p':
			writepid = 1;
			break;
		case 'N':
			/* NOP */
			break;
		case 'r':
			retries = atoi(optarg);
			if (retries <= 0 &&
			    retries != -1 && strcmp(optarg, "0") != 0) {
				if (!quiet)
					fprintf(stderr, "dotlockfile: "
						"-r %s: invalid argument\n",
						optarg);
				return L_ERROR;
			}
			if (retries == -1) {
				/* 4000 years */
				retries = 0x7ffffff0;
			}
			break;
		case 'm':
			if ((pwd = getpwuid(geteuid())) == NULL) {
				if (!quiet)
					fprintf(stderr, "dotlockfile: You don't exist. Go away.\n");
				return L_ERROR;
			}
			lockfile = mlockname(pwd->pw_name);
			if (!lockfile) {
				if (!quiet)
					perror("dotlockfile");
				return L_ERROR;
			}
			break;
		case 'l':
			lock = 1;
			break;
		case 'u':
			unlock = 1;
			break;
		case 'c':
			check = 1;
			break;
		case 'i':
			interval = atoi(optarg);
			if (interval <= 0 && strcmp(optarg, "0") != 0) {
				fprintf(stderr, "dotlockfile: -i needs argument >= 0\n");
				return L_ERROR;
			}
			flags |= L_INTERVAL;
			args.interval = interval;
			break;
		case 't':
			touch = 1;
			break;
		case 'P':
			passthrough = 1;
			break;
		default:
			usage();
			break;
	}

	/*
	 * next argument may be lockfile name
	 */
	if (!lockfile) {
		if (optind == argc)
			usage();
		lockfile = argv[optind++];
	}

	/*
	 * next arguments may be command [args...]
	 */
	if (optind < argc)
		cmd = argv + optind;

	/*
	 *	Options sanity check
	 */
	if ((cmd || lock) && (touch || check || unlock))
		usage();

	if (writepid)
		flags |= (cmd ? L_PID : L_PPID);

#ifdef MAXPATHLEN
	if (strlen(lockfile) >= MAXPATHLEN) {
		if (!quiet)
			fprintf(stderr, "dotlockfile: %s: name too long\n", lockfile);
		return L_NAMELEN;
	}
#endif

	/*
	 *	Check if we run setgid.
	 */
	int cwd_fd = -1;
	int need_privs = 0;
#ifdef MAILGROUP
	if (gid != egid) {
		/*
		 *	See if the requested lock is for a mailbox.
		 *	First, remember currect working directory.
		 */
#ifdef O_PATH
		cwd_fd = open(".", O_PATH|O_CLOEXEC);
#else
		cwd_fd = open(".", O_RDONLY|O_CLOEXEC);
#endif
		if (cwd_fd < 0) {
			if (!quiet)
				fprintf(stderr, "dotlockfile: opening \".\": %s\n",
					strerror(errno));
			return L_ERROR;
		}
		/*
		 *	Now change directory to the directory the lockfile is in.
		 */
		char *file, *dir;
		r = fn_split(lockfile, &file, &dir);
		if (r != L_SUCCESS) {
			if (!quiet)
				perror("dotlockfile");
			return L_ERROR;
		}
		if (chdir(dir) != 0) {
			if (!quiet)
				fprintf(stderr, "dotlockfile: %s: %s\n", dir, strerror(errno));
			return L_ERROR;
		}

		lockfile = file;
		need_privs = is_maillock(lockfile);
	}
#endif

	/*
	 *	See if we actually need to run setgid.
	 */
	if (need_privs) {
		if (setregid(gid, egid) != 0)
			perror_exit("setregid");
	} else {
		if (gid != egid && setgid(gid) != 0)
			perror_exit("setgid");
	}

	/*
	 *	Simple check for a valid lockfile ?
	 */
	if (check)
		return (lockfile_check(lockfile, flags) < 0) ? 1 : 0;


	/*
	 *	Touch lock ?
	 */
	if (touch)
		return (lockfile_touch(lockfile) < 0) ? 1 : 0;

	/*
	 *	Remove lockfile?
	 */
	if (unlock)
		return (lockfile_remove(lockfile) == 0) ? 0 : 1;


	/*
	 *	No, lock.
	 */
	r = lockfile_create_set_tmplock(lockfile, &tmplock, retries, flags, &args);
	if (r != 0 || !cmd)
		return r;


	/*
	 *	Spawn command.
	 *
	 *	Using an empty signal handler means that we ignore the
	 *	signal, but that it's restored to SIG_DFL at execve().
	 */
	set_signal(SIGINT, ignore_signal);
	set_signal(SIGQUIT, ignore_signal);
	set_signal(SIGHUP, ignore_signal);
	set_signal(SIGALRM, ignore_signal);

	pid_t pid = fork();
	if (pid < 0) {
		if (!quiet)
			perror("fork");
		lockfile_remove(lockfile);
		exit(L_ERROR);
	}
	if (pid == 0) {
		/* drop setgid */
		if (gid != egid && setgid(gid) < 0) {
			perror("setgid");
			exit(127);
		}
		/* restore current working directory */
		if (cwd_fd >= 0) {
			if (fchdir(cwd_fd) < 0) {
				perror("dotlockfile: restoring cwd:");
				exit(127);
			}
			close(cwd_fd);
		}
		/* exec */
		execvp(cmd[0], cmd);
		perror(cmd[0]);
		exit(127);
	}

	/* wait for child */
	int e, wstatus;
	while (1) {
		if (!writepid)
			alarm(30);
		e = waitpid(pid, &wstatus, 0);
		if (e >= 0 || errno != EINTR)
			break;
		if (!writepid)
			lockfile_touch(lockfile);
	}

	alarm(0);
	lockfile_remove(lockfile);

	if (passthrough) {
		if (WIFEXITED(wstatus))
			return WEXITSTATUS(wstatus);
		if (WIFSIGNALED(wstatus))
			return 128+WTERMSIG(wstatus);
	}
	return 0;
}

